### PR TITLE
perf(ui): render streamed markdown incrementally

### PR DESCRIPTION
**Pull request by marius-kilocode**
Long streamed responses repeatedly parsed, sanitized, rebuilt, and diffed the entire accumulated Markdown document for every visible update. The cost grows with response length even though earlier top-level blocks are already complete and cannot change.

This change splits streaming Markdown into stable completed blocks and one mutable tail. Completed blocks retain stable cache keys and mounted DOM nodes, while only the active tail is reparsed and replaced. Reference-style Markdown and unfinished fenced code retain the conservative existing path, and completion still performs the canonical full render. Sanitization, syntax highlighting, Mermaid rendering, copy controls, and streaming cadence remain unchanged.

The DOM implementation and block splitting live in Kilo-owned files. Shared upstream files contain only the narrow hooks needed to return block metadata and invoke the incremental renderer, with `kilocode_change` annotations on every changed line.

### Why this is worth doing

Markdown rendering becomes more important as responses and sessions grow. With the original renderer, every streamed update revisits the full accumulated response, so work increases throughout a long answer even though almost all earlier blocks are already final. In a large session, that repeated parsing, sanitization, temporary DOM creation, and reconciliation also competes with the retained transcript, tool cards, timeline, scrolling, and background session updates on the same renderer thread.

This optimization changes the growing part of that workload from the full response to the active Markdown tail. The benefit therefore compounds where it matters most: long answers, documentation-heavy work, code reviews, plans, and mature sessions with substantial transcript DOM. It is not a substitute for future transcript virtualization, but it removes avoidable per-update work before that retained-DOM cost is addressed. The reduction in transient nodes is particularly valuable because it lowers allocation and garbage-collection pressure rather than only moving CPU work elsewhere.

### Profile results

The benchmark replayed the same Agent Manager workload for each implementation: 13,531 Markdown characters in 677 foreground chunks, two concurrent background sessions, and 1,353 total part updates. Every run settled to the same 540-element Markdown DOM and canonical completed HTML.

| Metric | Original renderer | Stable block cache | Incremental tail DOM | Improvement vs original |
|---|---:|---:|---:|---:|
| Chromium script duration | 733.6 ms | 569.1 ms | 468.1 ms median | 36.2% lower |
| HTML parsing | 73.9 ms | 45.1 ms | 14.5 ms median | 80.4% lower |
| Transient Chromium nodes | 34,779 | 11,621 | 3,229 median | 90.7% lower |
| Trace script time | 3,951.8 ms | 3,142.8 ms | 2,900.8 ms median | 26.6% lower |
| Style recalculation | 193.7 ms | 141.7 ms | 183.1 ms median | 5.5% lower |
| Layout | 87.4 ms | 74.8 ms | 84.1 ms median | 3.8% lower |

The final incremental implementation was profiled twice. Chromium script duration measured 462.7 ms and 473.4 ms, while HTML parsing measured 14.7 ms and 14.3 ms. The reported final values are the medians of those two confirmation runs.

### CODE REVIEW

File: packages/ui/src/kilocode/markdown-incremental-dom.ts [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    const root = document.createElement("div")
+    root.innerHTML = html
+    decorate(root, labels)
+    const fragment = document.createDocumentFragment()
+    while (root.firstChild) fragment.appendChild(root.firstChild)
+    return fragment
+  }
+
+  const remove = (record: Record) => {
+    let node: ChildNode | null = record.start
```
[the remark this answers is not in the document]
Comment (marius-kilocode) [reply packages/ui/src/kilocode/markdown-incremental-dom.ts:48]: Good catch. I changed removal to first verify both boundaries share a parent and collect the complete bounded range before mutating anything. If the end marker is missing or unreachable, it now aborts and falls back to a clean rebuild instead of deleting adjacent records. I also added a regression test for a removed end boundary.

**Review by imanolmzd-svg — APPROVED** [review]

File: packages/kilo-vscode/tests/markdown-incremental-dom.spec.ts [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+    return {
+      updated,
+      tags: Array.from(container.children).map((child) => child.tagName),
+      text: container.textContent,
+    }
+  })
+
+  expect(result).toEqual({ updated: true, tags: ["P", "P"], text: "StableMiddle" })
+})
```
[the remark this answers is not in the document]
Comment (marius-kilocode) [reply packages/kilo-vscode/tests/markdown-incremental-dom.spec.ts:208]: Agreed. I restored all three cases in the Chromium spec: unsupported single/all-live blocks fall back without swallowing the render, lifecycle hooks run only for a handled streaming update, and an unchanged live tail can be promoted while a wrapper-free next tail is appended. The focused Playwright suite now has all five original scenarios and passes.
